src/main.rs

```diff
@@ -94,14 +96,15 @@ enum Format {
     If(Expr, Box<Format>, Box<Format>),
 }
 
-#[derive(Debug)]
-struct Lookahead {
-    pattern: Vec<ByteSet>,
+#[derive(Clone, Debug)]
+struct Switch {
+    accept: Option<usize>,
+    branches: Vec<(ByteSet, Switch)>,
 }
```

Comment:
Is this sort of like a tree of lookahead? Keeping track of the index of variant matched in the `accept` field?

Reply:
Yep! If none of the (disjoint) bytesets match or if there is no input remaining then the accept field indicates which if any of the switch cases succeeds.